README
NAME
    indirect - Lexically warn about using the indirect method call syntax.

VERSION
    Version 0.33

SYNOPSIS
    In a script :

        no indirect;               # lexically enables the pragma
        my $x = new Apple 1, 2, 3; # warns
        {
         use indirect;     # lexically disables the pragma
         my $y = new Pear; # legit, does not warn
         {
          # lexically specify an hook called for each indirect construct
          no indirect hook => sub {
           die "You really wanted $_[0]\->$_[1] at $_[2]:$_[3]"
          };
          my $z = new Pineapple 'fresh'; # croaks 'You really wanted...'
         }
        }
        try { ... }; # warns if try() hasn't been declared in this package

        no indirect 'fatal';     # or ':fatal', 'FATAL', ':Fatal' ...
        if (defied $foo) { ... } # croaks, note the typo

    Global uses :

        # Globally enable the pragma from the command-line
        perl -M-indirect=global -e 'my $x = new Banana;' # warns

        # Globally enforce the pragma each time perl is executed
        export PERL5OPT="-M-indirect=global,fatal"
        perl -e 'my $y = new Coconut;' # croaks

DESCRIPTION
    When enabled, this pragma warns about indirect method calls that are
    present in your code.

    The indirect syntax is now considered harmful, since its parsing has
    many quirks and its use is error prone : when the subroutine "foo" has
    not been declared in the current package, "foo $x" actually compiles to
    "$x->foo", and "foo { key => 1 }" to "'key'->foo(1)". In
    <http://www.shadowcat.co.uk/blog/matt-s-trout/indirect-but-still-fatal>,
    Matt S. Trout gives an example of an undesirable indirect method call on
    a block that can cause a particularly bewildering error.

    This pragma currently does not warn for core functions ("print", "say",
    "exec" or "system"). This may change in the future, or may be added as
    optional features that would be enabled by passing options to
    "unimport".

    This module is not a source filter.

METHODS
  "unimport"
        no indirect;
        no indirect 'fatal';
        no indirect hook => sub { my ($obj, $name, $file, $line) = @_; ... };
        no indirect 'global';
        no indirect 'global, 'fatal';
        no indirect 'global', hook => sub { ... };

    Magically called when "no indirect @opts" is encountered. Turns the
    module on. The policy to apply depends on what is first found in @opts :

    *   If it is a string that matches "/^:?fatal$/i", the compilation will
        croak when the first indirect method call is found.

        This option is mutually exclusive with the 'hook' option.

    *   If the key/value pair "hook => $hook" comes first, $hook will be
        called for each error with a string representation of the object as
        $_[0], the method name as $_[1], the current file as $_[2] and the
        line number as $_[3]. If and only if the object is actually a block,
        $_[0] is assured to start by '{'.

        This option is mutually exclusive with the 'fatal' option.

    *   If none of "fatal" and "hook" are specified, a warning will be
        emitted for each indirect method call.

    *   If @opts contains a string that matches "/^:?global$/i", the pragma
        will be globally enabled for all code compiled after the current "no
        indirect" statement, except for code that is in the lexical scope of
        "use indirect". This option may come indifferently before or after
        the "fatal" or "hook" options, in the case they are also passed to
        "unimport".

        The global policy applied is the one resulting of the "fatal" or
        "hook" options, thus defaults to a warning when none of those are
        specified :

            no indirect 'global';                # warn for any indirect call
            no indirect qw<global fatal>;        # die on any indirect call
            no indirect 'global', hook => \&hook # custom global action

        Note that if another policy is installed by a "no indirect"
        statement further in the code, it will overrule the global policy :

            no indirect 'global';  # warn globally
            {
             no indirect 'fatal';  # throw exceptions for this lexical scope
             ...
             require Some::Module; # the global policy will apply for the
                                   # compilation phase of this module
            }

  "import"
        use indirect;

    Magically called at each "use indirect". Turns the module off.

    As explained in "unimport"'s description, an "use indirect" statement
    will lexically override a global policy previously installed by "no
    indirect 'global', ..." (if there's one).

FUNCTIONS
  "msg"
        my $msg = msg($object, $method, $file, $line);

    Returns the default error message that "indirect" generates when an
    indirect method call is reported.

CONSTANTS
  "I_THREADSAFE"
    True iff the module could have been built with thread-safety features
    enabled.

  "I_FORKSAFE"
    True iff this module could have been built with fork-safety features
    enabled. This will always be true except on Windows where it's false for
    perl 5.10.0 and below .

DIAGNOSTICS
  "Indirect call of method "%s" on object "%s" at %s line %d."
    The default warning/exception message thrown when an indirect method
    call on an object is found.

  "Indirect call of method "%s" on a block at %s line %d."
    The default warning/exception message thrown when an indirect method
    call on a block is found.

ENVIRONMENT
  "PERL_INDIRECT_PM_DISABLE"
    If this environment variable is set to true when the pragma is used for
    the first time, the XS code won't be loaded and, although the 'indirect'
    lexical hint will be set to true in the scope of use, the pragma itself
    won't do anything. In this case, the pragma will always be considered to
    be thread-safe, and as such "I_THREADSAFE" will be true. This is useful
    for disabling "indirect" in production environments.

    Note that clearing this variable after "indirect" was loaded has no
    effect. If you want to re-enable the pragma later, you also need to
    reload it by deleting the 'indirect.pm' entry from %INC.

CAVEATS
    The implementation was tweaked to work around several limitations of
    vanilla "perl" pragmas : it's thread safe, and does not suffer from a
    "perl 5.8.x-5.10.0" bug that causes all pragmas to propagate into
    "require"d scopes.

    Before "perl" 5.12, "meth $obj" (no semicolon) at the end of a file is
    not seen as an indirect method call, although it is as soon as there is
    another token before the end (as in "meth $obj;" or "meth $obj 1"). If
    you use "perl" 5.12 or greater, those constructs are correctly reported.

    With 5.8 perls, the pragma does not propagate into "eval STRING". This
    is due to a shortcoming in the way perl handles the hints hash, which is
    addressed in perl 5.10.

    Indirect constructs that appear in code "eval"'d during the global
    destruction phase of a spawned thread or pseudo-fork (the processes used
    internally for the "fork" emulation on Windows) are not reported.

    The search for indirect method calls happens before constant folding.
    Hence "my $x = new Class if 0" will be caught.

DEPENDENCIES
    perl 5.8.1.

    A C compiler. This module may happen to build with a C++ compiler as
    well, but don't rely on it, as no guarantee is made in this regard.

    Carp (standard since perl 5), XSLoader (since perl 5.6.0).

AUTHOR
    Vincent Pit, "<perl at profvince.com>", <http://www.profvince.com>.

    You can contact me by mail or on "irc.perl.org" (vincent).

BUGS
    Please report any bugs or feature requests to "bug-indirect at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=indirect>. I will be
    notified, and then you'll automatically be notified of progress on your
    bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc indirect

    Tests code coverage report is available at
    <http://www.profvince.com/perl/cover/indirect>.

ACKNOWLEDGEMENTS
    Bram, for motivation and advices.

    Andrew Main and Florian Ragwitz, for testing on real-life code and
    reporting issues.

COPYRIGHT & LICENSE
    Copyright 2008,2009,2010,2011,2012,2013,2014 Vincent Pit, all rights
    reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for indirect

0.33    2014-09-29 20:20 UTC
        + Fix : [RT #99083] : Breaks eval in an END block in Win32 pseudo-forks.
                Thanks Graham Knop for reporting.
        + Fix : Segfaults during global destruction of a thread or a
                pseudo-fork.

0.32    2014-09-21 20:15 UTC
        + Add : Support for the PERL_OP_PARENT optional feature introduced in
                perl 5.21.2.
        + Fix : [RT #92806] : Tests that use run_perl() fail on Android
                Thanks Brian Fraser for the patch.
        + Fix : indirect constructs will no longer segfault while inside
                the empty package on perl 5.8.x. This fix may also prevent
                some segfaults during global destruction.

0.31    2013-09-05 16:45 UTC
        + Fix : [RT #88428] : no indirect in eval can trigger for direct calls
                on __PACKAGE__
                Thanks Graham Knop for reporting.
        + Tst : Author tests are no longer bundled with this distribution.
                They are only made available to authors in the git repository.

0.30    2013-05-16 15:55 UTC
        + Fix : [RT #83806] : false positives with Devel::Declare
                [RT #83839] : false positive using ? : syntax
                Thanks Andrew Main for the patch.
                However, please note that the reason this patch seems to fix
                thinks has not been explained.
        + Fix : [RT #84649] : incorrect RT link in metadata
                Thanks Karen Etheridge for reporting.

0.29    2013-03-05 01:30 UTC
        + Fix : [RT #83659] : false positives
                Proper method calls in string-like environments (like
                "@{[ $x->new ]}" will no longer be reported as indirect.
                This was a regression in 0.28.
                Thanks Andrew Main for reporting.
        + Fix : Broken linkage on Windows with gcc 3.4, which appears in
                particular when using ActivePerl's default compiler suite.
                For those setups, the indirect shared library will now be
                linked against the perl dll directly (instead of the import
                library).

0.28    2013-02-26 17:05 UTC
        + Fix : [RT #83450] : newlines confuse indirect
                Perl sometimes resets the line buffer between the object and
                the method name (e.g. for "sort Class\n->method" outside of
                eval), and this could cause direct method calls to be reported
                as indirect.
                Thanks Gianni Ceccarelli for reporting.
        + Fix : Check functions are now replaced and restored in a thread-safe
                manner, either by using the wrap_op_checker() function from perl
                when it is available (starting from perl 5.16) or by taking the
                OP_REFCNT mutex on older perls.

0.27    2013-01-30 19:00 UTC
        + Fix : [RT #82562] : indirect/Devel::CallParser interaction
                indirect has been taught to play nicely with Devel::CallParser.
                Thanks Andrew Main for the patch.
        + Tst : Author tests overhaul.

0.26    2011-10-23 14:25 UTC
        + Add : "no indirect 'global'" enables the pragma for the whole program,
                except for lexical scopes that "use indirect" explicitely.
        + Chg : Passing both the 'fatal' and 'hook' options to unimport() is
                now considered an error, and will result in an exception.
                unimport() used to consider only the first passed option of
                those two, and silently ignored the other.
        + Tst : Test failures of t/41-threads-teardown.t and t/50-external.t on
                Cygwin should have been addressed.
        + Tst : Threads tests will not fail anymore if resources constraints
                prevent the system from creating all the required threads.

0.25    2011-08-24 15:40 UTC
        + Fix : RT #69291 is now also fixed for perl 5.8. The pragma will no
                longer vivify the "indirect" entry in the hints hash %^H on
                perl 5.8.
        + Tst : Attempt to make t/50-external.t pass on Cygwin.

0.24    2011-07-17 23:15 UTC
        + Fix : [RT #64521] : "no indirect" leaking into eval.
                This is currently only fixed for perl 5.10 (perl 5.12 and
                higher were never affected). It was caused by a very stupid
                mistake of mine that was introduced in indirect version 0.23.
                Thanks Michael G Schwern for reporting.
        + Fix : [RT #69291] : indirect.pm breaks %^H.
                This was caused by the same mistake as for the previous bug,
                and as such it is also only fixed for perl 5.10 (and never
                affected perl 5.12).
                Thanks Andrew Main for reporting.
        + Doc : C++ compilers are officially NOT supported.

0.23    2010-10-03 00:15 UTC
        + Fix : Some indirect constructs could be incorrectly reported when
                several modules were used in the same scope. This caused
                t/30-scope.t to fail randomly.
        + Tst : Threads tests are now only run on perl 5.13.4 and higher.
                They could segfault randomly because of what seems to be an
                internal bug of Perl, which has been addressed in 5.13.4.
                There is also an environment variable that allows you to
                forcefully run those tests, but it should be set only for
                author testing and not for end users.

0.22    2010-08-16 16:00 UTC
        + Add : Indirect constructs are now reported for code interpolated
                in quote-like environments, like "${\( ... )}", "@{[ ... ]}",
                s/pattern/ ... /e, qr/(?{ ... })/ or qr/(??{ ... })/.
        + Add : You can now make the pragma lethal by passing anything matching
                /^:?fatal$/i to import(), including "FATAL" and ":Fatal".

t/lib/indirect/TestThreads.pm
package indirect::TestThreads;

use strict;
use warnings;

use Config qw<%Config>;

use VPIT::TestHelpers;

sub import {
 shift;

 require indirect;

 skip_all 'This indirect isn\'t thread safe' unless indirect::I_THREADSAFE();

 my $force = $ENV{PERL_INDIRECT_TEST_THREADS} ? 1 : !1;
 skip_all 'This perl wasn\'t built to support threads'
                                                    unless $Config{useithreads};
 skip_all 'perl 5.13.4 required to test thread safety'
                                              unless $force or "$]" >= 5.013004;

 load_or_skip_all('threads', $force ? '0' : '1.67', [ ]);

 my %exports = (
  spawn => \&spawn,
 );

 my $pkg = caller;
 while (my ($name, $code) = each %exports) {
  no strict 'refs';
  *{$pkg.'::'.$name} = $code;
 }
}

sub spawn {
 local $@;
 my @diag;
 my $thread = eval {
  local $SIG{__WARN__} = sub { push @diag, "Thread creation warning: @_" };
  threads->create(@_);
 };
 push @diag, "Thread creation error: $@" if $@;
 if (@diag) {
  require Test::Leaner;
  Test::Leaner::diag($_) for @diag;
 }
 return $thread ? $thread : ();
}

1;


ptable.h
/* This file is part of the indirect Perl module.
 * See http://search.cpan.org/dist/indirect/ */

/* This is a pointer table implementation essentially copied from the ptr_table
 * implementation in perl's sv.c, except that it has been modified to use memory
 * shared across threads.
 * Copyright goes to the original authors, bug reports to me. */

/* This header is designed to be included several times with different
 * definitions for PTABLE_NAME and PTABLE_VAL_FREE(). */

#undef pPTBLMS
#undef pPTBLMS_
#undef aPTBLMS
#undef aPTBLMS_

/* Context for PerlMemShared_* functions */

#ifdef PERL_IMPLICIT_SYS
# define pPTBLMS  pTHX
# define pPTBLMS_ pTHX_
# define aPTBLMS  aTHX
# define aPTBLMS_ aTHX_
#else
# define pPTBLMS
# define pPTBLMS_
# define aPTBLMS
# define aPTBLMS_
#endif

#ifndef pPTBL
# define pPTBL  pPTBLMS
#endif
#ifndef pPTBL_
# define pPTBL_ pPTBLMS_
#endif
#ifndef aPTBL
# define aPTBL  aPTBLMS
#endif
#ifndef aPTBL_
# define aPTBL_ aPTBLMS_
#endif

#ifndef PTABLE_NAME
# define PTABLE_NAME ptable
#endif

#ifndef PTABLE_VAL_FREE
# define PTABLE_VAL_FREE(V)
#endif

#ifndef PTABLE_JOIN
# define PTABLE_PASTE(A, B) A ## B
# define PTABLE_JOIN(A, B)  PTABLE_PASTE(A, B)
#endif

#ifndef PTABLE_PREFIX
# define PTABLE_PREFIX(X) PTABLE_JOIN(PTABLE_NAME, X)
#endif

#ifndef ptable_ent
typedef struct ptable_ent {
 struct ptable_ent *next;
 const void *       key;
 void *             val;
} ptable_ent;
#define ptable_ent ptable_ent
#endif /* !ptable_ent */

#ifndef ptable
typedef struct ptable {
 ptable_ent **ary;
 size_t       max;
 size_t       items;
} ptable;
#define ptable ptable
#endif /* !ptable */

#ifndef ptable_new
STATIC ptable *ptable_new(pPTBLMS) {
#define ptable_new() ptable_new(aPTBLMS)
 ptable *t = PerlMemShared_malloc(sizeof *t);
 t->max   = 15;
 t->items = 0;
 t->ary   = PerlMemShared_calloc(t->max + 1, sizeof *t->ary);
 return t;
}
#endif /* !ptable_new */

#ifndef PTABLE_HASH
# define PTABLE_HASH(ptr) \
     ((PTR2UV(ptr) >> 3) ^ (PTR2UV(ptr) >> (3 + 7)) ^ (PTR2UV(ptr) >> (3 + 17)))
#endif

#ifndef ptable_find
STATIC ptable_ent *ptable_find(const ptable * const t, const void * const key) {
#define ptable_find ptable_find
 ptable_ent *ent;
 const UV hash = PTABLE_HASH(key);

 ent = t->ary[hash & t->max];
 for (; ent; ent = ent->next) {
  if (ent->key == key)
   return ent;
 }

 return NULL;
}
#endif /* !ptable_find */

#ifndef ptable_fetch
STATIC void *ptable_fetch(const ptable * const t, const void * const key) {
#define ptable_fetch ptable_fetch
 const ptable_ent *const ent = ptable_find(t, key);

 return ent ? ent->val : NULL;
}
#endif /* !ptable_fetch */

#ifndef ptable_split
STATIC void ptable_split(pPTBLMS_ ptable * const t) {
#define ptable_split(T) ptable_split(aPTBLMS_ (T))
 ptable_ent **ary = t->ary;
 const size_t oldsize = t->max + 1;
 size_t newsize = oldsize * 2;
 size_t i;

 ary = PerlMemShared_realloc(ary, newsize * sizeof(*ary));
 Zero(&ary[oldsize], newsize - oldsize, sizeof(*ary));
 t->max = --newsize;
 t->ary = ary;

 for (i = 0; i < oldsize; i++, ary++) {
  ptable_ent **curentp, **entp, *ent;
  if (!*ary)
   continue;
  curentp = ary + oldsize;
  for (entp = ary, ent = *ary; ent; ent = *entp) {
   if ((newsize & PTABLE_HASH(ent->key)) != i) {
    *entp     = ent->next;
    ent->next = *curentp;
    *curentp  = ent;
    continue;
   } else
    entp = &ent->next;
  }
 }
}
#endif /* !ptable_split */

STATIC void PTABLE_PREFIX(_store)(pPTBL_ ptable * const t, const void * const key, void * const val) {
 ptable_ent *ent = ptable_find(t, key);

 if (ent) {
  void *oldval = ent->val;
  PTABLE_VAL_FREE(oldval);
  ent->val = val;
 } else if (val) {
  const size_t i = PTABLE_HASH(key) & t->max;
  ent = PerlMemShared_malloc(sizeof *ent);
  ent->key  = key;
  ent->val  = val;
  ent->next = t->ary[i];
  t->ary[i] = ent;
  t->items++;
  if (ent->next && t->items > t->max)
   ptable_split(t);
 }
}

#ifndef ptable_walk
STATIC void ptable_walk(pTHX_ ptable * const t, void (*cb)(pTHX_ ptable_ent *ent, void *userdata), void *userdata) {
#define ptable_walk(T, CB, UD) ptable_walk(aTHX_ (T), (CB), (UD))
 if (t && t->items) {
  register ptable_ent ** const array = t->ary;
  size_t i = t->max;
  do {
   ptable_ent *entry;
   for (entry = array[i]; entry; entry = entry->next)
    cb(aTHX_ entry, userdata);
  } while (i--);
 }
}
#endif /* !ptable_walk */

STATIC void PTABLE_PREFIX(_clear)(pPTBL_ ptable * const t) {
 if (t && t->items) {
  register ptable_ent ** const array = t->ary;
  size_t i = t->max;

  do {
   ptable_ent *entry = array[i];
   while (entry) {
    ptable_ent * const oentry = entry;
    void *val = oentry->val;
    entry = entry->next;
    PTABLE_VAL_FREE(val);
    PerlMemShared_free(oentry);
   }
   array[i] = NULL;
  } while (i--);

  t->items = 0;
 }
}

STATIC void PTABLE_PREFIX(_free)(pPTBL_ ptable * const t) {
 if (!t)
  return;
 PTABLE_PREFIX(_clear)(aPTBL_ t);
 PerlMemShared_free(t->ary);
 PerlMemShared_free(t);
}

#undef pPTBL
#undef pPTBL_
#undef aPTBL
#undef aPTBL_

#undef PTABLE_NAME
#undef PTABLE_VAL_FREE


Makefile.PL
use 5.008001;

use strict;
use warnings;
use ExtUtils::MakeMaker;

use Config;

my @DEFINES;
my %macro;

my $is_gcc_34 = 0;
print "Checking if this is gcc 3.4 on Windows trying to link against an import library... ";
if ($^O eq 'MSWin32' and not grep /^LD[A-Z]*=/, @ARGV) {
 my ($libperl, $gccversion) = map $_ || '', @Config{qw<libperl gccversion>};
 if ($gccversion =~ /^3\.4\.[0-9]+/ and $libperl =~ s/\.lib$//) {
  $is_gcc_34 = 1;
  my ($lddlflags, $ldflags) = @Config{qw<lddlflags ldflags>};
  $_ ||= '', s/-L(?:".*?"|\S+)//g for $lddlflags, $ldflags;
  $libperl = "-l$libperl";
  my $libdirs = join ' ',
                 map { s/(?<!\\)((?:\\\\)*")/\\$1/g; qq[-L"$_"] }
                  @Config{qw<bin sitebin>};
  $macro{LDDLFLAGS}    = "$lddlflags $libdirs $libperl";
  $macro{LDFLAGS}      = "$ldflags $libdirs $libperl";
  $macro{PERL_ARCHIVE} = '',
 }
}
print $is_gcc_34 ? "yes\n" : "no\n";

# Threads, Windows and 5.8.x don't seem to be best friends
if ($^O eq 'MSWin32' and "$]" < 5.009) {
 push @DEFINES, '-DI_MULTIPLICITY=0';
}

# Fork emulation got "fixed" in 5.10.1
if ($^O eq 'MSWin32' and "$]" < 5.010_001) {
 push @DEFINES, '-DI_FORKSAFE=0';
}

@DEFINES = (DEFINE => join ' ', @DEFINES) if @DEFINES;
%macro   = (macro  => { %macro })         if %macro; # Beware of the circle

my $dist = 'indirect';

(my $name = $dist) =~ s{-}{::}g;

(my $file = $dist) =~ s{-}{/}g;
$file = "lib/$file.pm";

my %PREREQ_PM = (
 'Carp'     => 0,
 'XSLoader' => 0,
);

my %BUILD_REQUIRES =(
 'Config'              => 0,
 'ExtUtils::MakeMaker' => 0,
 'Test::More'          => 0,
 %PREREQ_PM,
);

my %META = (
 configure_requires => {
  'Config'              => 0,
  'ExtUtils::MakeMaker' => 0,
 },
 build_requires => {
  %BUILD_REQUIRES,
 },
 dynamic_config => 1,
 resources => {
  bugtracker => "http://rt.cpan.org/Dist/Display.html?Name=$dist",
  homepage   => "http://search.cpan.org/dist/$dist/",
  license    => 'http://dev.perl.org/licenses/',
  repository => "http://git.profvince.com/?p=perl%2Fmodules%2F$dist.git",
 },
);

WriteMakefile(
 NAME             => $name,
 AUTHOR           => 'Vincent Pit <perl@profvince.com>',
 LICENSE          => 'perl',
 VERSION_FROM     => $file,
 ABSTRACT_FROM    => $file,
 PL_FILES         => {},
 @DEFINES,
 BUILD_REQUIRES   => \%BUILD_REQUIRES,
 PREREQ_PM        => \%PREREQ_PM,
 MIN_PERL_VERSION => '5.008001',
 META_MERGE       => \%META,
 dist             => {
  PREOP    => "pod2text -u $file > \$(DISTVNAME)/README",
  COMPRESS => 'gzip -9f', SUFFIX => 'gz'
 },
 clean            => {
  FILES => "$dist-* *.gcov *.gcda *.gcno cover_db Debian_CPANTS.txt*"
 },
 %macro,
);


lib/indirect.pm
package indirect;

use 5.008001;

use strict;
use warnings;

=head1 NAME

indirect - Lexically warn about using the indirect method call syntax.

=head1 VERSION

Version 0.30

=cut

our $VERSION;
BEGIN {
 $VERSION = '0.30';
}

=head1 SYNOPSIS

In a script :

    no indirect;               # lexically enables the pragma
    my $x = new Apple 1, 2, 3; # warns
    {
     use indirect;     # lexically disables the pragma
     my $y = new Pear; # legit, does not warn
     {
      # lexically specify an hook called for each indirect construct
      no indirect hook => sub {
       die "You really wanted $_[0]\->$_[1] at $_[2]:$_[3]"
      };
      my $z = new Pineapple 'fresh'; # croaks 'You really wanted...'
     }
    }
    try { ... }; # warns if try() hasn't been declared in this package

    no indirect 'fatal';     # or ':fatal', 'FATAL', ':Fatal' ...
    if (defied $foo) { ... } # croaks, note the typo

Global uses :

    # Globally enable the pragma from the command-line
    perl -M-indirect=global -e 'my $x = new Banana;' # warns

    # Globally enforce the pragma each time perl is executed
    export PERL5OPT="-M-indirect=global,fatal"
    perl -e 'my $y = new Coconut;' # croaks

=head1 DESCRIPTION

When enabled, this pragma warns about indirect method calls that are present in your code.

The indirect syntax is now considered harmful, since its parsing has many quirks and its use is error prone : when the subroutine C<foo> has not been declared in the current package, C<foo $x> actually compiles to C<< $x->foo >>, and C<< foo { key => 1 } >> to C<< 'key'->foo(1) >>.
In L<http://www.shadowcat.co.uk/blog/matt-s-trout/indirect-but-still-fatal>, Matt S. Trout gives an example of an undesirable indirect method call on a block that can cause a particularly bewildering error.

This pragma currently does not warn for core functions (C<print>, C<say>, C<exec> or C<system>).
This may change in the future, or may be added as optional features that would be enabled by passing options to C<unimport>.

This module is B<not> a source filter.

=cut

BEGIN {
 if ($ENV{PERL_INDIRECT_PM_DISABLE}) {
  *_tag = sub ($) { 1 };
  *I_THREADSAFE = sub () { 1 };
  *I_FORKSAFE   = sub () { 1 };
 } else {
  require XSLoader;
  XSLoader::load(__PACKAGE__, $VERSION);
 }
}

=head1 METHODS

=head2 C<unimport>

    no indirect;
    no indirect 'fatal';
    no indirect hook => sub { my ($obj, $name, $file, $line) = @_; ... };
    no indirect 'global';
    no indirect 'global, 'fatal';
    no indirect 'global', hook => sub { ... };

Magically called when C<no indirect @opts> is encountered.
Turns the module on.
The policy to apply depends on what is first found in C<@opts> :

=over 4

=item *

If it is a string that matches C</^:?fatal$/i>, the compilation will croak when the first indirect method call is found.

This option is mutually exclusive with the C<'hook'> option.

=item *

If the key/value pair C<< hook => $hook >> comes first, C<$hook> will be called for each error with a string representation of the object as C<$_[0]>, the method name as C<$_[1]>, the current file as C<$_[2]> and the line number as C<$_[3]>.
If and only if the object is actually a block, C<$_[0]> is assured to start by C<'{'>.

This option is mutually exclusive with the C<'fatal'> option.

=item *

If none of C<fatal> and C<hook> are specified, a warning will be emitted for each indirect method call.

=item *

If C<@opts> contains a string that matches C</^:?global$/i>, the pragma will be globally enabled for B<all> code compiled after the current C<no indirect> statement, except for code that is in the lexical scope of C<use indirect>.
This option may come indifferently before or after the C<fatal> or C<hook> options, in the case they are also passed to L</unimport>.

The global policy applied is the one resulting of the C<fatal> or C<hook> options, thus defaults to a warning when none of those are specified :

    no indirect 'global';                # warn for any indirect call
    no indirect qw<global fatal>;        # die on any indirect call
    no indirect 'global', hook => \&hook # custom global action

Note that if another policy is installed by a C<no indirect> statement further in the code, it will overrule the global policy :

    no indirect 'global';  # warn globally
    {
     no indirect 'fatal';  # throw exceptions for this lexical scope
     ...
     require Some::Module; # the global policy will apply for the
                           # compilation phase of this module
    }

=back

=cut

sub _no_hook_and_fatal {
 require Carp;
 Carp::croak("The 'fatal' and 'hook' options are mutually exclusive");
}

sub unimport {
 shift;

 my ($global, $fatal, $hook);

 while (@_) {
  my $arg = shift;
  if ($arg eq 'hook') {
   _no_hook_and_fatal() if $fatal;
   $hook = shift;
  } elsif ($arg =~ /^:?fatal$/i) {
   _no_hook_and_fatal() if defined $hook;
   $fatal = 1;
  } elsif ($arg =~ /^:?global$/i) {
   $global = 1;
  }
 }

 unless (defined $hook) {
  $hook = $fatal ? sub { die msg(@_) } : sub { warn msg(@_) };
 }

 $^H |= 0x00020000;
 if ($global) {
  delete $^H{+(__PACKAGE__)};
  _global($hook);
 } else {
  $^H{+(__PACKAGE__)} = _tag($hook);
 }

 return;
}

=head2 C<import>

    use indirect;

Magically called at each C<use indirect>. Turns the module off.

As explained in L</unimport>'s description, an C<use indirect> statement will lexically override a global policy previously installed by C<no indirect 'global', ...> (if there's one).

=cut

sub import {
 $^H |= 0x00020000;
 $^H{+(__PACKAGE__)} = _tag(undef);

 return;
}

=head1 FUNCTIONS

=head2 C<msg>

    my $msg = msg($object, $method, $file, $line);

Returns the default error message that C<indirect> generates when an indirect method call is reported.

=cut

sub msg {
 my $obj = $_[0];

 join ' ', "Indirect call of method \"$_[1]\" on",
           ($obj =~ /^\s*\{/ ? "a block" : "object \"$obj\""),
           "at $_[2] line $_[3].\n";
};

=head1 CONSTANTS

=head2 C<I_THREADSAFE>

True iff the module could have been built with thread-safety features enabled.

=head2 C<I_FORKSAFE>

True iff this module could have been built with fork-safety features enabled.
This will always be true except on Windows where it's false for perl 5.10.0 and below .

=head1 DIAGNOSTICS

=head2 C<Indirect call of method "%s" on object "%s" at %s line %d.>

The default warning/exception message thrown when an indirect method call on an object is found.

=head2 C<Indirect call of method "%s" on a block at %s line %d.>

The default warning/exception message thrown when an indirect method call on a block is found.

=head1 ENVIRONMENT

=head2 C<PERL_INDIRECT_PM_DISABLE>

If this environment variable is set to true when the pragma is used for the first time, the XS code won't be loaded and, although the C<'indirect'> lexical hint will be set to true in the scope of use, the pragma itself won't do anything.
In this case, the pragma will always be considered to be thread-safe, and as such L</I_THREADSAFE> will be true.
This is useful for disabling C<indirect> in production environments.

Note that clearing this variable after C<indirect> was loaded has no effect.
If you want to re-enable the pragma later, you also need to reload it by deleting the C<'indirect.pm'> entry from C<%INC>.

=head1 CAVEATS

The implementation was tweaked to work around several limitations of vanilla C<perl> pragmas : it's thread safe, and does not suffer from a C<perl 5.8.x-5.10.0> bug that causes all pragmas to propagate into C<require>d scopes.

Before C<perl> 5.12, C<meth $obj> (no semicolon) at the end of a file is not seen as an indirect method call, although it is as soon as there is another token before the end (as in C<meth $obj;> or C<meth $obj 1>).
If you use C<perl> 5.12 or greater, those constructs are correctly reported.

With 5.8 perls, the pragma does not propagate into C<eval STRING>.
This is due to a shortcoming in the way perl handles the hints hash, which is addressed in perl 5.10.

The search for indirect method calls happens before constant folding.
Hence C<my $x = new Class if 0> will be caught.

=head1 DEPENDENCIES

L<perl> 5.8.1.

A C compiler.
This module may happen to build with a C++ compiler as well, but don't rely on it, as no guarantee is made in this regard.

L<Carp> (standard since perl 5), L<XSLoader> (since perl 5.006).

=head1 AUTHOR

Vincent Pit, C<< <perl at profvince.com> >>, L<http://www.profvince.com>.

You can contact me by mail or on C<irc.perl.org> (vincent).

=head1 BUGS

Please report any bugs or feature requests to C<bug-indirect at rt.cpan.org>, or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=indirect>.
I will be notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc indirect

Tests code coverage report is available at L<http://www.profvince.com/perl/cover/indirect>.

=head1 ACKNOWLEDGEMENTS

Bram, for motivation and advices.

Andrew Main and Florian Ragwitz, for testing on real-life code and reporting issues.

=head1 COPYRIGHT & LICENSE

Copyright 2008,2009,2010,2011,2012,2013 Vincent Pit, all rights reserved.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut

1; # End of indirect
